package pomagma_messaging;

message Histogram {
  message Term {
    required uint64 count = 1;
    optional uint32 ob = 2;
    optional string name = 3;
  }
  repeated Term terms = 1;
}

message AnalystRequest {

  message TestInference {
  }
  message Simplify {
    repeated string codes = 1;
  }
  message Validate {
    repeated string codes = 1;
  }
  message ValidateCorpus {
    message Line {
      optional string name = 1;
      required string code = 2;
    }
    repeated Line lines = 1;
  }
  message GetHistogram {
  }
  message FitLanguage {
    optional Histogram histogram = 1;
  }

  optional string id = 1;
  optional TestInference test_inference = 2;
  optional Simplify simplify = 3;
  optional Validate validate = 4;
  optional ValidateCorpus validate_corpus = 5;
  optional GetHistogram get_histogram = 6;
  optional FitLanguage fit_language = 7;
}

message AnalystResponse {
  enum Trool {
    MAYBE = 0;
    FALSE = 1;
    TRUE = 2;
  }
  message Validity {
    required Trool is_top = 1 [default = MAYBE];
    required Trool is_bot = 2 [default = MAYBE];
    required bool pending = 3 [default = false];
  }

  message TestInference {
    required uint64 fail_count = 1;
  }
  message Simplify {
    repeated string codes = 1;
  }
  message Validate {
    repeated Validity results = 1;
  }
  message ValidateCorpus {
    repeated Validity results = 1;
  }
  message GetHistogram {
    required Histogram histogram = 1;
  }
  message FitLanguage {
    message Symbol {
      required string name = 1;
      required float prob = 2;
    }
    repeated Symbol symbols = 1;
  }

  optional string id = 1;
  optional TestInference test_inference = 2;
  optional Simplify simplify = 3;
  optional Validate validate = 4;
  optional ValidateCorpus validate_corpus = 5;
  optional GetHistogram get_histogram = 6;
  optional FitLanguage fit_language = 7;

  repeated string error_log = 8;
}
